cpp/src/io/orc/orc_reader.cu

```diff
@@ -0,0 +1,731 @@
+
+  // Update chunks with pointers to column data
+  for (size_t i = 0; i < num_stripes; ++i) {
+    for (size_t j = 0; j < num_columns; ++j) {
+      auto &chunk = chunks[i * num_columns + j];
+      chunk.valid_map_base = reinterpret_cast<uint32_t *>(columns[j]->valid);
+      chunk.column_data_base = columns[j]->data;
+      chunk.dtype_len = (columns[j]->dtype == GDF_STRING)
+                            ? sizeof(std::pair<const char *, size_t>)
+                            : gdf_dtype_size(columns[j]->dtype);
+    }
+  }
+
+  // Allocate global dictionary for deserializing
+  rmm::device_vector<orc::gpu::DictionaryEntry> global_dict(num_dicts);
+
+  CUDA_TRY(cudaMemcpyAsync(chunks.device_ptr(), chunks.host_ptr(),
+                           chunks.memory_size(), cudaMemcpyHostToDevice));
+  CUDA_TRY(DecodeNullsAndStringDictionaries(
+      chunks.device_ptr(), global_dict.data().get(), num_columns, num_stripes,
+      num_rows, skip_rows));
+  CUDA_TRY(DecodeOrcColumnData(chunks.device_ptr(), global_dict.data().get(),
+                               num_columns, num_stripes, num_rows, skip_rows));
+  CUDA_TRY(cudaMemcpyAsync(chunks.host_ptr(), chunks.device_ptr(),
+                           chunks.memory_size(), cudaMemcpyDeviceToHost));
+  CUDA_TRY(cudaStreamSynchronize(0));
+
+  LOG_PRINTF("[+] Decoded Column Information\n");
+  for (size_t i = 0; i < num_columns; ++i) {
+    for (size_t j = 0; j < num_stripes; ++j) {
+      columns[i]->null_count += chunks[j * num_columns + i].null_count;
+    }
+    LOG_PRINTF(
+        "columns[%zd].null_count = %d/%d (start_row=%d, nrows=%d, "
+        "strm_len=%d)\n",
+        i, columns[i]->null_count, columns[i]->size, chunks[i].start_row,
+        chunks[i].num_rows, chunks[i].strm_len[orc::gpu::CI_PRESENT]);
+  }
+}
+
+/**
+ * @brief Reads Apache ORC data and returns an array of gdf_columns.
+ *
+ * @param[in,out] args Structure containing input and output args
+ *
+ * @return gdf_error GDF_SUCCESS if successful, otherwise an error code.
+ **/
+gdf_error read_orc(orc_read_arg *args) {
+
+  DataSource input(args->source);
+
+  OrcMetadata md(input.data(), input.size());
+  CUDF_EXPECTS(md.get_num_columns() > 0, "No columns found");
+
+  // Select only required stripes (aka rowgroups)
+  const auto stripes = md.select_stripes(args->skip_rows, args->num_rows);
+  const int num_rows = md.get_total_rows();
+
+  // Select only columns required (if it exists), otherwise select all
+  std::vector<int32_t> selected_cols;
+  if (args->use_cols) {
+    std::vector<std::string> use_names(args->use_cols,
+                                       args->use_cols + args->use_cols_len);
+    int index = 0;
+    for (const auto &use_name : use_names) {
+      for (int i = 0; i < md.get_num_columns(); ++i, ++index) {
+        if (index >= md.get_num_columns()) {
+          index = 0;
+        }
+        if (md.ff.GetColumnName(index) == use_name) {
+          selected_cols.emplace_back(index);
+          index++;
```

Comment:
Missing break ?

Reply:
Hrmm true... I guess we can't get duplicate column names in ORC?

Reply:
I'd imagine we'd want to still pick only the first matching one if there were (I was mainly concerned about perf here, like in the GFN dataset with 7000+ columns)

Reply:
Okay, fixed.